community/index/src/main/java/org/neo4j/index/internal/gbptree/CrashGenCleaner.java

```diff
@@ -0,0 +1,245 @@
+            }
+        };
+    }
+
+    // === Methods about checking if a tree node has crashed pointers ===
+
+    private boolean hasCrashedGSPP( TreeNode<?,?> treeNode, PageCursor cursor ) throws IOException
+    {
+        boolean isTreeNode;
+        int keyCount;
+        do
+        {
+            isTreeNode = TreeNode.nodeType( cursor ) == TreeNode.NODE_TYPE_TREE_NODE;
+            keyCount = treeNode.keyCount( cursor );
+        }
+        while ( cursor.shouldRetry() );
+
+        if ( !isTreeNode )
+        {
+            return false;
+        }
+
+        boolean hasCrashed;
+        do
+        {
+            hasCrashed =
+                    hasCrashedGSPP( cursor, TreeNode.BYTE_POS_NEWGEN ) ||
+                    hasCrashedGSPP( cursor, TreeNode.BYTE_POS_LEFTSIBLING ) ||
+                    hasCrashedGSPP( cursor, TreeNode.BYTE_POS_RIGHTSIBLING );
+
+            if ( !hasCrashed && TreeNode.isInternal( cursor ) )
+            {
+                for ( int i = 0; i <= keyCount && i <= maxKeyCount && !hasCrashed; i++ )
+                {
+                    hasCrashed |= hasCrashedGSPP( cursor, treeNode.childOffset( i ) );
+                }
+            }
+        }
+        while ( cursor.shouldRetry() );
+        return hasCrashed;
+    }
+
+    private boolean hasCrashedGSPP( PageCursor cursor, int gsppOffset )
+    {
+        return hasCrashedGSP( cursor, gsppOffset ) || hasCrashedGSP( cursor, gsppOffset + GenSafePointer.SIZE );
+    }
+
+    private boolean hasCrashedGSP( PageCursor cursor, int offset )
+    {
+        cursor.setOffset( offset );
+        long generation = GenSafePointer.readGeneration( cursor );
+        return generation > stableGeneration && generation < unstableGeneration;
+    }
+
+    // === Methods about actually cleaning a discovered crashed tree node ===
+
+    private void cleanTreeNode( TreeNode<?,?> treeNode, PageCursor cursor, AtomicInteger cleanedPointers )
+    {
+        cleanCrashedGSPP( cursor, TreeNode.BYTE_POS_NEWGEN, cleanedPointers );
+        cleanCrashedGSPP( cursor, TreeNode.BYTE_POS_LEFTSIBLING, cleanedPointers );
+        cleanCrashedGSPP( cursor, TreeNode.BYTE_POS_RIGHTSIBLING, cleanedPointers );
+
+        if ( TreeNode.isInternal( cursor ) )
+        {
+            int keyCount = treeNode.keyCount( cursor );
+            for ( int i = 0; i <= keyCount && i <= maxKeyCount; i++ )
+            {
+                cleanCrashedGSPP( cursor, treeNode.childOffset( i ), cleanedPointers );
+            }
+        }
+    }
+
+    private void cleanCrashedGSPP( PageCursor cursor, int gsppOffset, AtomicInteger cleanedPointers )
+    {
+        cleanCrashedGSP( cursor, gsppOffset, cleanedPointers );
+        cleanCrashedGSP( cursor, gsppOffset + GenSafePointer.SIZE, cleanedPointers );
+    }
+
+    private void cleanCrashedGSP( PageCursor cursor, int gspOffset, AtomicInteger cleanedPointers )
+    {
+        if ( hasCrashedGSP( cursor, gspOffset ) )
```

Comment:
lets add a comment about that, for future maintainers to know about that assumption